### PR TITLE
Add basic multi node test

### PR DESCRIPTION
**Pull request by dapplion**
This PR is open to showcase the test case were with more than 1 node, no checkpoint is justified. To run it
```
cd packages/lodestar
yarn test:e2e
```

To edit the combinations of nodes to achieve justification edit this line
https://github.com/ChainSafe/lodestar/blob/937da487136327a081beb6e35b36c94ff0c27b66/packages/lodestar/test/e2e/noEth1Sim.test.ts#L20

----

Also, when trying to clean up the nodes and validators I'm not able to do it without the afterEach hook failing with
```
2020-07-22 16:34:09  []                error: uncaughtException: Pubsub has not started
Error: Pubsub has not started
    at LodestarGossipsub.unsubscribe (/home/lion/Code/eth2.0/lodestar/node_modules/libp2p-gossipsub/src/pubsub.js:335:19)
    at /home/lion/Code/eth2.0/lodestar/packages/lodestar/src/network/gossip/gossip.ts:88:29
    at Array.forEach (<anonymous>)
    at Gossip.unsubscribe (/home/lion/Code/eth2.0/lodestar/packages/lodestar/src/network/gossip/gossip.ts:87:37)
    at Gossip.unsubscribeFromAttestationSubnet (/home/lion/Code/eth2.0/lodestar/packages/lodestar/src/network/gossip/gossip.ts:83:14)
    at Timeout.AttestationCollector.unsubscribeSubnet [as _onTimeout] (/home/lion/Code/eth2.0/lodestar/packages/lodestar/src/sync/utils/attestation-collector.ts:22:33)
    at listOnTimeout (internal/timers.js:533:17)
    at processTimers (internal/timers.js:475:7) 

```
or
```
2020-07-22 18:42:27  [VALIDATOR 0-7]   error: Failed to produce attestation slot=24, committee=0, 
                        error=Cannot read property 'state' of null
```

It seams that `.stop()` resolves before everything is cleaned up too, so there are still logs being printed from the previous node set

### CODE REVIEW

**Comment by mpetrunic** [comment]
@wemeetagain You mentioned you know what's causing Pubsub has not started ? ^^

@dapplion I don't think you need to run metrics at all in tests? I suspect metrics are initialized twice on same port or something

**Comment by dapplion** [comment]
@mpetrunic How can I disable the metrics programmatically?

**Comment by wemeetagain** [comment]
RE metrics error
Pass in `{ metrics: { enabled: false } }` as options to the `BeaconNode` to disable the metrics server.

But it doesn't look like this is a problem with the metrics server, rather a problem with the metrics themselves (which are not disable-able).
We're using [`prom-client`](https://github.com/siimon/prom-client#default-metrics) to collect default metrics from the nodejs instance. Every metric is registered with a Registry. (I guess, judging from the error, that a metric can only be registered once). 
Is the node being `start`ed multiple times?

**Comment by dapplion** [comment]
@wemeetagain Yep, most likely the nodes where being re-started. I've scoped the node instances to the `it` instance and now the error doesn't happen. :+1:

**Comment by dapplion** [comment]
When running > 1 node I see that almost all calls to aggregate and proof fail with 
```
2020-07-22 18:47:58  [VALIDATOR 0-1]   error: Failed to produce aggregate and proof Error: No matching attestations found for attestationData
Error: No matching attestations found for attestationData
    at ValidatorApi.produceAggregateAndProof (/home/lion/Code/eth2.0/lodestar/packages/lodestar/src/api/impl/validator/validator.ts:112:19)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:94:5)
    at Timeout._onTimeout (/home/lion/Code/eth2.0/lodestar/packages/lodestar-validator/src/services/attestation.ts:192:27)
```
May this cause the aggregate attention to not be submitted at all and never contribute to justification?

---

Note that I added this error
https://github.com/ChainSafe/lodestar/blob/ecd6c9d807dd0742cdf483d62713187bdbf436aa/packages/lodestar/src/api/impl/validator/validator.ts#L195-L197
otherwise, it throws a TypeError for trying to reduce an empty array without initial value.

**Comment by mpetrunic** [comment]
This would indicate that validator tried to aggregate on different node from where he published attestation. That said , other nice should have received attestation via gossip.

But more likely is there is some bug with fetching or supporting requires attestations

File: packages/lodestar/test/e2e/noEth1Sim.test.ts [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+  const validatorsPerNode = 8;
+  const beaconParams: Partial<IBeaconParams> = {
+    SECONDS_PER_SLOT: 2,
+    SLOTS_PER_EPOCH: 8
+  };
+
+  let onDoneHandlers: (() => Promise<void>)[] = [];
+
+  for (const nodeCount of [1, 2]) {
```
Comment (mpetrunic) [inline packages/lodestar/test/e2e/noEth1Sim.test.ts]: we already have test for single node :)

You should call multinode test or something

**Comment by wemeetagain** [comment]
file conflict

**Comment by dapplion** [comment]
At least locally, it still doesn't justify any checkpoint for up to 10 minutes.

**Comment by dapplion** [comment]
The CI e2e test logs are spammed with a ton of errors, they are barely readable. This test should not be merged until the underlying issues are resolved or it would add 20 minutes of runtime + a ton of unrelated data to other fronts of work.

**Comment by dapplion** [comment]
I'll add below relevant errors from the last e2e test run:

---

The error that actually stopped the tests is not a timeout but `MaxBufferError`:
```
2020-07-27T18:24:20.9046863Z lerna ERR! MaxBufferError: stdout maxBuffer exceeded
2020-07-27T18:24:20.9048325Z lerna ERR!     at PassThrough.<anonymous> (/home/runner/work/lodestar/lodestar/node_modules/get-stream/index.js:41:19)
2020-07-27T18:24:20.9050475Z lerna ERR!     at PassThrough.emit (events.js:327:22)
2020-07-27T18:24:20.9051970Z lerna ERR!     at addChunk (_stream_readable.js:295:12)
2020-07-27T18:24:20.9053568Z lerna ERR!     at readableAddChunk (_stream_readable.js:267:11)
2020-07-27T18:24:20.9054975Z lerna ERR!     at PassThrough.Readable.push (_stream_readable.js:212:10)
2020-07-27T18:24:20.9056520Z lerna ERR!     at PassThrough.Transform.push (_stream_transform.js:152:32)
2020-07-27T18:24:20.9057895Z lerna ERR!     at PassThrough.afterTransform (_stream_transform.js:96:10)
2020-07-27T18:24:20.9059762Z lerna ERR!     at PassThrough._transform (_stream_passthrough.js:46:3)
2020-07-27T18:24:20.9061135Z lerna ERR!     at PassThrough.Transform._read (_stream_transform.js:191:10)
2020-07-27T18:24:20.9093540Z lerna WARN complete Waiting for 1 child process to exit. CTRL-C to exit immediately.
2020-07-27T18:24:20.9097722Z lerna ERR! lerna stdout maxBuffer exceeded
2020-07-27T18:24:20.9157820Z ##[error]Process completed with exit code 1.
```

Had this ever happened before? I had no idea lerna had a log limit and didn't just stream to process

---

There seems to be a significant difference in epoch timing between validators and nodes, with this error showing up for almost all validators
```
2020-07-27T18:24:20.8898717Z @chainsafe/lodestar: 2020-07-27 18:24:20  [VALIDATOR 3-0]   [31merror[39m: Failed to obtain proposer duties Error: Cannot get duties for epoch more than two ahead: 205 <= 162
2020-07-27T18:24:20.8899541Z @chainsafe/lodestar: Error: Cannot get duties for epoch more than two ahead: 205 <= 162
2020-07-27T18:24:20.8900240Z @chainsafe/lodestar:     at Object.lte (/home/runner/work/lodestar/lodestar/packages/lodestar-utils/src/assert.ts:31:13)
2020-07-27T18:24:20.8900663Z @chainsafe/lodestar:     at ValidatorApi.getProposerDuties (/home/runner/work/lodestar/lodestar/packages/lodestar/src/api/impl/validator/validator.ts:63:33)
2020-07-27T18:24:20.8901583Z @chainsafe/lodestar:     at runMicrotasks (<anonymous>)
2020-07-27T18:24:20.8901974Z @chainsafe/lodestar:     at runNextTicks (internal/process/task_queues.js:62:5)
2020-07-27T18:24:20.8902324Z @chainsafe/lodestar:     at listOnTimeout (internal/timers.js:518:9)
2020-07-27T18:24:20.8903252Z @chainsafe/lodestar:     at processTimers (internal/timers.js:492:7)
2020-07-27T18:24:20.8903943Z @chainsafe/lodestar:     at /home/runner/work/lodestar/lodestar/packages/lodestar-validator/src/services/block.ts:64:24
```
The same error happen for multiple validators between 18:24:16 and 18:24:18, where each validator is requesting duties for completely different epochs:
- VALIDATOR 3-0: 206
- VALIDATOR 3-2: 195
- VALIDATOR 1-5: 192
- VALIDATOR 1-7: 223

---

Many attempts to destructure an nullish variable

```
2020-07-27T18:24:08.7911426Z @chainsafe/lodestar: 2020-07-27 18:24:08  [VALIDATOR 0-1]   [31merror[39m: Failed to obtain attester duty for epoch 192 Error: Cannot destructure property 'epochCtx' of '(intermediate value)' as it is null.
2020-07-27T18:24:08.7912087Z @chainsafe/lodestar: TypeError: Cannot destructure property 'epochCtx' of '(intermediate value)' as it is null.
2020-07-27T18:24:08.7912469Z @chainsafe/lodestar:     at ValidatorApi.getAttesterDuties (/home/runner/work/lodestar/lodestar/packages/lodestar/src/api/impl/validator/validator.ts:76:17)
2020-07-27T18:24:08.7912813Z @chainsafe/lodestar:     at runMicrotasks (<anonymous>)
2020-07-27T18:24:08.7913124Z @chainsafe/lodestar:     at runNextTicks (internal/process/task_queues.js:62:5)
2020-07-27T18:24:08.7913441Z @chainsafe/lodestar:     at listOnTimeout (internal/timers.js:518:9)
2020-07-27T18:24:08.7913757Z @chainsafe/lodestar:     at processTimers (internal/timers.js:492:7)
2020-07-27T18:24:08.7914389Z @chainsafe/lodestar:     at /home/runner/work/lodestar/lodestar/packages/lodestar-validator/src/services/attestation.ts:74:24
```
```
2020-07-27T18:24:08.7906784Z @chainsafe/lodestar: 2020-07-27 18:24:08  [VALIDATOR 0-1]   [31merror[39m: Failed to obtain proposer duties Error: Cannot destructure property 'state' of '(intermediate value)' as it is null.
2020-07-27T18:24:08.7907822Z @chainsafe/lodestar: TypeError: Cannot destructure property 'state' of '(intermediate value)' as it is null.
2020-07-27T18:24:08.7908576Z @chainsafe/lodestar:     at ValidatorApi.getProposerDuties (/home/runner/work/lodestar/lodestar/packages/lodestar/src/api/impl/validator/validator.ts:61:17)
2020-07-27T18:24:08.7909035Z @chainsafe/lodestar:     at runMicrotasks (<anonymous>)
2020-07-27T18:24:08.7909351Z @chainsafe/lodestar:     at runNextTicks (internal/process/task_queues.js:62:5)
2020-07-27T18:24:08.7909677Z @chainsafe/lodestar:     at listOnTimeout (internal/timers.js:518:9)
2020-07-27T18:24:08.7909991Z @chainsafe/lodestar:     at processTimers (internal/timers.js:492:7)
2020-07-27T18:24:08.7910638Z @chainsafe/lodestar:     at /home/runner/work/lodestar/lodestar/packages/lodestar-validator/src/services/block.ts:64:24
```

---

Others:

```
2020-07-27T17:26:41.4778288Z @chainsafe/lodestar: 2020-07-27 17:26:41  [VALIDATOR 1-5]   [31merror[39m: Failed to subscribe to committee subnet Error: Cannot collect attestations before regular sync
2020-07-27T17:26:41.4778883Z @chainsafe/lodestar: Error: Cannot collect attestations before regular sync
2020-07-27T17:26:41.4779295Z @chainsafe/lodestar:     at BeaconSync.collectAttestations (/home/runner/work/lodestar/lodestar/packages/lodestar/src/sync/sync.ts:97:19)
2020-07-27T17:26:41.4779705Z @chainsafe/lodestar:     at ValidatorApi.subscribeCommitteeSubnet (/home/runner/work/lodestar/lodestar/packages/lodestar/src/api/impl/validator/validator.ts:147:25)
2020-07-27T17:26:41.4780080Z @chainsafe/lodestar:     at runMicrotasks (<anonymous>)
2020-07-27T17:26:41.4780434Z @chainsafe/lodestar:     at runNextTicks (internal/process/task_queues.js:62:5)
2020-07-27T17:26:41.4780791Z @chainsafe/lodestar:     at listOnTimeout (internal/timers.js:518:9)
2020-07-27T17:26:41.4781144Z @chainsafe/lodestar:     at processTimers (internal/timers.js:492:7)
2020-07-27T17:26:41.4781809Z @chainsafe/lodestar:     at /home/runner/work/lodestar/lodestar/packages/lodestar-validator/src/services/attestation.ts:99:11
```

**Comment by wemeetagain** [comment]
re: validators being on different epochs, my immediate thought is that we're probably DOSing ourselves in host of ways. The CPU gets pushed to 100%, and some nodes don't get updated properly in time, likely requests get responded to after many seconds, etc, etc. I'd like to try again once #1233 and #1235 get landed, which will help alleviate a few of the slow processing codepaths.

Though, even after those PRs land, there's still one big bottleneck of creating a slotSignature and subscribing to a subnet, which take 2+seconds for a single aggregator, and happen simultaneously for all aggregators. If we're running that at the same time on all 4 nodes, that alone will throw off the slot times.

If you'd like to test this theory, you can run the tests with the inspector turned on, and profile. Likely won't be pretty.

**Comment by wemeetagain** [comment]
Ok, I've done some profiling, several wildly incorrect statements in my last comment:
07-27-4node
(pic is showing 4 node test w/ modified slot time of 10s)


It's taking my computer ~2s for _all_ subscriptions to subnets, not for 1

Also, looking at these flame charts, they're already mostly dominated by bls operations. The speedups of #1233 or #1235 will likely not do much.

**Comment by mpetrunic** [comment]
Have you tried merging master? I remember some of those stuff should be fixed on master

**Comment by mpetrunic** [comment]
Also, we should probably run each node in it's own thread. We cannot expect 4 nodes to run in single thread

**Comment by mpetrunic** [comment]
Seems like 4 bns are too much for ci.

I think we should first move 3 nodes to separate threads (https://nodejs.org/docs/latest-v12.x/api/worker_threads.html) and have 4th one  in main thread where we can monitor justification event

**Review by mpetrunic — APPROVED** [review]

**Review by wemeetagain — APPROVED** [review]

**Pull request merged**